padatious/bracket_expansion.py

```diff
@@ -0,0 +1,129 @@
+# Copyright 2017 Mycroft AI, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+class Fragment(object):
+    """(Abstract) empty sentence fragment"""
+
+    def __init__(self, tree):
+        self._tree = tree
+
+    def expand(self):
+        """Expanded version of the fragment."""
+        # Creates one empty sentence
+        return [[]]
+
+    def __str__(self):
+        return self._tree.__str__()
+
+    def __repr__(self):
+        return self._tree.__repr__()
+
+
+class Word(Fragment):
+    """Single word in the sentence tree."""
+
+    def expand(self):
+        # Creates one sentence that contains exactly that word
+        return [[self._tree]]
+
+
+class Sentence(Fragment):
+    """A Sentence made of several concatenations/words."""
+
+    def expand(self):
+        # Creates a combination of all sub-sentences
+        old_expanded = [[]]
+        for sub in self._tree:
+            sub_expanded = sub.expand()
+            new_expanded = []
+            while len(old_expanded) > 0:
+                sentence = old_expanded.pop()
+                for new in sub_expanded:
+                    new_expanded.append(sentence + new)
+            old_expanded = new_expanded
+        return old_expanded
+
+
+class Options(Fragment):
+    """A Combination of possible sub-sentences."""
+
+    def expand(self):
+        # Returns all of its options as seperated sub-sentences

```

Comment:
Could you this and the other comments above a docstring so that the auto generated docs will have more info?

Reply:
Done!